File: SortingAlgos/MergeSort/python/MergeSort.py
# merge fn
def merge(List, start, end, mid):
	temp = []
	i = start
	j = mid+1

	# checking & adding smallest number to temp
	while i<=mid and j<=end:
		if List[i] <= List[j]:
			temp.append(List[i])
			# k+=1
			i+=1
		else:
			temp.append(List[j])
			# k+=1
			j+=1
		
	# adding rest from left subList to temp
	while i<=mid:
		temp.append(List[i])
		# k+=1
		i+=1
	
	# adding rest from right subList to temp
	while j<=end:
		temp.append(List[j])
		# k+=1
		j+=1

	# adding items from temp to List
	for i in range(start,end+1):
		List[i] = temp[i-start]

# mergeSort fn
def mergeSort(List, start, end):
	if start < end:
		mid = int((start+end)/2)
		mergeSort(List, start, mid)
		mergeSort(List, mid+1, end)
		merge(List, start, end, mid)
	

File: SortingAlgos/MergeSort/cpp/MergeSort.cpp
#include <iostream>
using namespace std;

// merge splitted single elemented arrays into one array
void merge(int array[], int start, int end, int mid)
{
	// creating temporary array to store sorted items; of size end-start+1
	int temp[end - start + 1];
	int i = start, j = mid + 1, k = 0;

	// checking & adding smallest item to temp
	while (i <= mid && j <= end)
	{
		if (array[i] <= array[j])
		{
			temp[k] = array[i];
			k++;
			i++;
		}
		else
		{
			temp[k] = array[j];
			k++;
			j++;
		}
	}

	// adding rest of the left array to temp
	while (i <= mid)
	{
		temp[k] = array[i];
		k++;
		i++;
	}

	// adding rest of the right array to temp
	while (j <= end)
	{
		temp[k] = array[j];
		k++;
		j++;
	}

	// to add items in temp to array
	for (int i = start; i <= end; i++)
	{
		array[i] = temp[i - start];
	}
}

// splits the input array into array with single elements; using recursion
void mergeSort(int array[], int start, int end)
{
	if (start < end)
	{
		int mid = (start + end) / 2;
		// left half of array
		mergeSort(array, start, mid);
		// right half of array
		mergeSort(array, mid + 1, end);
		merge(array, start, end, mid);
	}
}
